File: frontend/src/stores/beliefs.js
import { ref, computed } from 'vue'
import { defineStore } from 'pinia'
import { beliefsApi } from '../api/beliefs'

export const useBeliefsStore = defineStore('beliefs', () => {
  const rounds = ref([])
  const dimensions = ref([])
  const mode = ref(null) // 'keyword' | 'llm' | 'demo'
  const loading = ref(false)
  const error = ref(null)

  const hasData = computed(() => rounds.value.length > 0)

  const latestRound = computed(() =>
    rounds.value.length ? rounds.value[rounds.value.length - 1] : null,
  )

  async function fetchDimensions() {
    try {
      const res = await beliefsApi.getDimensions()
      dimensions.value = res.data.dimensions
    } catch (e) {
      dimensions.value = fallbackDimensions()
    }
  }

  async function extractBeliefs(simulationId, actions, useLlm = false) {
    loading.value = true
    error.value = null
    try {
      const res = await beliefsApi.extract(simulationId, actions, useLlm)
      rounds.value = res.data.rounds
      mode.value = res.data.mode
    } catch (e) {
      error.value = e.message || 'Failed to extract beliefs'
    } finally {
      loading.value = false
    }
  }

  async function fetchDemo(numRounds = 10) {
    loading.value = true
    error.value = null
    try {
      const res = await beliefsApi.demo(numRounds)
      rounds.value = res.data.rounds
      mode.value = 'demo'
    } catch (e) {
      error.value = e.message || 'Failed to fetch demo beliefs'
    } finally {
      loading.value = false
    }
  }

  function reset() {
    rounds.value = []
    mode.value = null
    loading.value = false
    error.value = null
  }

  return {
    rounds,
    dimensions,
    mode,
    loading,
    error,
    hasData,
    latestRound,
    fetchDimensions,
    extractBeliefs,
    fetchDemo,
    reset,
  }
})

function fallbackDimensions() {
  return [
    { key: 'product_quality', label: 'Product Quality', color: '#2068FF' },
    { key: 'pricing', label: 'Pricing Perception', color: '#ff5600' },
    { key: 'brand_trust', label: 'Brand Trust', color: '#009900' },
    { key: 'competitive_position', label: 'Competitive Position', color: '#AA00FF' },
    { key: 'adoption_intent', label: 'Adoption Intent', color: '#E67E00' },
  ]
}

File: backend/app/api/beliefs.py
"""
Beliefs API
Endpoints for agent belief system tracking and analysis.
"""

from flask import Blueprint, jsonify, request

from ..config import Config
from ..services.belief_tracker import (
    BELIEF_DIMENSIONS,
    extract_beliefs_from_actions,
    extract_beliefs_llm,
    generate_demo_beliefs,
)
from ..utils.logger import get_logger

logger = get_logger('mirofish.beliefs')

beliefs_bp = Blueprint('beliefs', __name__, url_prefix='/api/v1/beliefs')


@beliefs_bp.route('/dimensions', methods=['GET'])
def get_dimensions():
    """Return the list of tracked belief dimensions with display metadata."""
    meta = {
        'product_quality': {'label': 'Product Quality', 'color': '#2068FF'},
        'pricing': {'label': 'Pricing Perception', 'color': '#ff5600'},
        'brand_trust': {'label': 'Brand Trust', 'color': '#009900'},
        'competitive_position': {'label': 'Competitive Position', 'color': '#AA00FF'},
        'adoption_intent': {'label': 'Adoption Intent', 'color': '#E67E00'},
    }
    return jsonify({
        'dimensions': [
            {'key': d, **meta.get(d, {'label': d, 'color': '#888'})}
            for d in BELIEF_DIMENSIONS
        ]
    })


@beliefs_bp.route('/<simulation_id>/extract', methods=['POST'])
def extract_beliefs(simulation_id):
    """
    Extract beliefs from simulation actions.

    Expects JSON body:
      { "actions": [...], "use_llm": false }

    If actions are not provided, returns demo data.
    """
    data = request.get_json(silent=True) or {}
    actions = data.get('actions', [])
    use_llm = data.get('use_llm', False)

    if not actions:
        logger.info(f"No actions for {simulation_id}, returning demo beliefs")
        beliefs = generate_demo_beliefs()
        return jsonify({
            'simulation_id': simulation_id,
            'mode': 'demo',
            'rounds': _format_rounds(beliefs),
        })

    if use_llm and Config.LLM_API_KEY:
        try:
            from ..utils.llm_client import LLMClient
            client = LLMClient()
            beliefs = extract_beliefs_llm(actions, client)
            mode = 'llm'
        except Exception as e:
            logger.warning(f"LLM extraction failed: {e}, falling back to keywords")
            beliefs = extract_beliefs_from_actions(actions)
            mode = 'keyword'
    else:
        beliefs = extract_beliefs_from_actions(actions)
        mode = 'keyword'

    return jsonify({
        'simulation_id': simulation_id,
        'mode': mode,
        'rounds': _format_rounds(beliefs),
    })


@beliefs_bp.route('/demo', methods=['GET'])
def demo_beliefs():
    """Return demo belief data for testing/preview."""
    num_rounds = request.args.get('rounds', 10, type=int)
    num_rounds = max(1, min(50, num_rounds))
    beliefs = generate_demo_beliefs(num_rounds)
    return jsonify({
        'simulation_id': 'demo',
        'mode': 'demo',
        'rounds': _format_rounds(beliefs),
    })


def _format_rounds(beliefs: dict) -> list:
    """Convert internal round dict to sorted list for API response."""
    return [
        {
            'round': rn,
            **data,
        }
        for rn, data in sorted(beliefs.items())
    ]
